src/effgen.ml

```diff
@@ -591,6 +637,48 @@ module GeneratorsWithContext (Ctx : Context) = struct
     in
     [ (3, gen) ]
 
+  and tuple_intro_rules env t eff size =
+    match t with
+    | Tuple [] | Tuple [ _ ] -> failwith "tuple_intro_rules: tuple must have arity > 1"
+    | Tuple t_lst ->
+      let gen st =
+        let valid_tuple_arity = List.length t_lst in
+        let size' = size / valid_tuple_arity in
+        let exception Short_circuit in
+        match
+          List.map
+            (fun t ->
+              match term_gen_sized env t eff size' st with
+              | Some trm -> trm
+              | None -> raise Short_circuit)
+            t_lst
+        with
+        | exception Short_circuit -> None
+        | trm_lst -> Some (Constructor (t, TupleArity valid_tuple_arity, trm_lst, eff))
+      in
+      [ (3, gen) ]
+    | _ -> []
+
+  and tuple_elim_rules env goal_t eff size =
+    let open Syntax in
+    (* generate a new tuple that is then pattern matched and [goal_t] term is returned *)
+    let elim_new_tuple_rule st =
+      let arity = Gen.(2 -- 10) st in
+      let size' = size / arity in
+      let pattern_match_gen =
+        Gen.list_size (return arity) (type_gen size') >>= fun types_lst ->
+        (* add [goal_t] to [types_lst] so that its term can be returned in future *)
+        Gen.shuffle_l (goal_t :: types_lst) >>= fun tuple_elt_types ->
+        term_gen_sized env (Tuple tuple_elt_types) eff size' >>=? fun tuple st ->
+        let pat, env = pattern_of_term env tuple st in
+        match term_gen_sized env goal_t eff (sqrt size) st with
+        | Some term -> Some (PatternMatch (goal_t, tuple, [ (pat, term) ], eff))
+        | None -> None
+      in
+      pattern_match_gen st
```

Comment:
I'm not convinced by this implementation of the elimination rules.

- Nitpick: you can simplify `let pattern_match_gen = .. in pattern_match_gen` into just `..`

- You should not need to add `goal_t` to `tuple_elt_types` to generate elements of that type; doing this does not make it easier to find an element of type `goal_t`, it only pushes the demand to a tuple element recursively. I think that just *not* doing this should be fine (getting rid of lines 670-671 and using `types_lst` instead of `tuple_elt_types`).

- I think that `pattern_of_term` is not the right approach: there is no reason for the pattern we destruct here to be directed by the shape of the generated term. We could either always generate a tuple pattern with variables inside (the simplest approach), or have a pattern generator that is also directed by the type (if an element type is a tuple itself, we can sometimes generate a tuple pattern recursively). In the long run, there will be a type-directed pattern-generation algorithm (I think such an algorithm could even generate a list of patterns, rather than a single pattern), but for now maybe just outputting a flat pattern (with only variables as element patterns) would be fine -- it is more in line with standard "tuple elimination rules" in small languages.

Reply:
> Nitpick: you can simplify let pattern_match_gen = .. in pattern_match_gen into just ..

I would think naming this a little complex chunk of code as `pattern_match_gen` and then applying `st` makes it a little easier to skim/read the code, doesn't it? The code should become nice when we switch to 4.08 with more powerful `let` syntax.

> You should not need to add goal_t to tuple_elt_types to generate elements of that type; doing this does not make it easier to find an element of type goal_t

I understand that adding `goal_t` to the tuple types list does not make the generation easier, but I thought that in future we might want with prob 1/2 to either take one of the elements of the tuple, which now includes the term with the goal type, or generate a new term with the goal type. 

Say, the goal type is `int`. If the tuple includes an `int`, which it does, as we included it in the tuple, the generator can either generate a new `int` term to return or reuse one of the elements of the tuple that has type `int`. I think reusing elements can be nice. 

> it only pushes the demand to a tuple element recursively

I think it is not what you mean, but if you want to say that inserting the goal type into the types list will force the generator to generate a type similar to the current tuple type, note that the goal type for elimination rule is not (necessarily) a tuple type, as noted in the comment above. 

> not doing this should be fine

Yes, unless we want the feature I mentioned earlier: take an element of the tuple and return it. 

> have a pattern generator that is also directed by the type

Will replace `pattern_of_term` with `pattern_of_type`. That indeed sounds more reasonable.

Reply:
> I would think naming this a little complex chunk of code as pattern_match_gen and then applying st makes it a little easier to skim/read the code, doesn't it?

Indeed, I had somehow missed the `st` application at the end. This makes sense and I think the current state is reasonable, thanks!

Reply:
> Yes, unless we want the feature I mentioned earlier: take an element of the tuple and return it.

I'm not convinced that this feature is very important, to the point of biasing all tuple generations we make (which also runs the risk of covering less of the interesting/buggy behaviors). In the body part of the tuple elimination, the deconstructed tuple elements are available in the environment, and that should be enough to allow reuse when one of the tuple elements happens to be the goal type. (Our weights should be biased towards reusing variables when they exist at the right type in the context, instead of inventing a new term each time; this is a general aspect of the general that is useful in many other scenarios.) I think this should be frequent enough (that a tuple type happens to be the goal type) to have examples of this kind generated, without needing to make the generator look less natural.

(We could *test* this by adding a test that fails if the template `let (x1, .., xn) = e in xi` appears in the generated term, and checking that we indeed manage to observe failure.)